---
title: "Quality check of probe selection for TAD fusion events"
author: "Grace Hansen"
date: "10/05/2020"
output: pdf_document
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/midway/TAD_fusion/')
library(data.table)
library(tidyverse)
library(scales)
library(gridExtra)

########## For color manipulation of graph ###############
darken <- function(color, factor=1.2){
  col <- col2rgb(color)
  col <- col/factor
  col <- rgb(t(col), maxColorValue=255)
  col
}
lighten <- function(color, factor=0.8){
  col <- col2rgb(color)
  col <- col/factor
  for (i in 1:length(col)) { if ( col[i] > 255 ) { col[i] = 255 } }
  col <- rgb(t(col), maxColorValue=255)
  col
}
#########################################################
```

# Summary of pipeline

Details of the pipeline can be found in the Snakefile and associated files.
For the Rao et al. GM12878 in-situ datasets, I:

1. Converted the data available on GEO (*nodups.txt.gz) to hicsummary format
2. Made a HOMER tag dir from the *hicsummary file
3. Combined all tag dirs into one tag dir
4. Identified TADs and calculated DI on each chromosome using HOMER
5. Merged the data from all chromosomes
6. Selected TADs > 100kb that overlapped no other TADs
7. For each TAD, selected the 15 MboI fragments with the highest DI score and the top 15 fragments with the lowest DI score
  + (15+15)*~3000 = ~90,000 fragments representing the extreme ends of each TAD, which in normal conditions should be shielded from extra-TAD contacts
8. For each fragment, selected 120bp candidate probes from the beginning, middle, and end of each fragment. Selected the first fragment with unique mapping and good GC content.
  
## HiC statistics:

* **27 libraries**
* **5.009E9 total reads**
* **1kb resolution**
  
# TAD calling

Did my TAD calling work?

### TAD calling: descriptive statistics

``` {r TAD_general_stats,echo=FALSE}
tads<-fread("~/midway/TAD_fusion/results/merged_HIC.tad.2D.bed")
nTADs<-paste("Number of TADs (no filtering): ",nrow(tads),sep='')
mean_dist=paste("Mean length: ",round(mean(tads$V3-tads$V2)/1000,digits=2)," kb",sep='')
median_dist=paste("Median length: ",round(median(tads$V3-tads$V2)/1000,digits=2)," kb",sep='')
min_dist=paste("Min length: ",round(min(tads$V3-tads$V2)/1000,digits=2)," kb",sep='')
max_dist=paste("Max length: ",round(max(tads$V3-tads$V2)/1000,digits=2)," kb",sep='')

ggplot()+
  geom_histogram(aes(x=(tads$V3-tads$V2)/1000),bins=50,fill="steelblue4")+
  scale_x_continuous(name="TAD distance (kb)")+
  theme_minimal()+
  annotate(x=1000,y=300,"text",label=paste(nTADs,"TAD statistics:",mean_dist,median_dist,max_dist,min_dist,sep='\n'))+
  ggtitle("All called TADs")
```

### TAD calling: internal consistency

Do the TADs called make sense relative to loops and directionality index (DI) also identified by HOMER? Below is a visualization of the loop ends, TADs, and DI on the WashU browser:

![](DI_loops_TADs.jpg)

### TAD calling: similarity to Rao et al. TAD calling

Rao et al used Arrowhead, not HOMER to call TADs. They also allowed overlapping TADs, which I didn't. Here's how my TAD calling compares to theirs:

![](Rao_vs_HOMER_TADs.jpg)

# Directionality index within TADs

Does directionality index follow the expected pattern within selected TADs?

``` {r DI_TADs,echo=FALSE}
#Grab top 25 pos and top 25 neg DI fragments within each of the top N TADs
all_frags<-fread("results/MboI_frags_TADs_avgDI.txt")
colnames(all_frags)<-c("chr","start","stop","DI")

tads_frags<-matrix(ncol=7)
for (i in 1:nrow(tads)) {
  tad_frags<-all_frags[all_frags$chr==paste("chr",tads$V1[i],sep='') & all_frags$start>tads$V2[i] & all_frags$stop<tads$V3[i]] #Grab fragments within TAD
  tad_frags$scaled_DI<-scale(tad_frags$DI,center = TRUE) #Center and scale DI values within each TAD to aid visualization
  tad_frags$pos_TAD<-as.numeric(rownames(tad_frags))/nrow(tad_frags)
  tad_frags$tad<-i
  tads_frags<-rbind(tads_frags,tad_frags,use.names=FALSE)
}
tads_frags<-tads_frags[2:nrow(tads_frags),]
colnames(tads_frags)<-c("chr","start","stop","DI","scaled_DI","pos_TAD","TAD_index")

# Remap DI values to positions defined as proportion of TAD
tads_DI_scaled<-matrix(ncol=3)
fracs<-seq(0.02,1,0.02)
for (i in unique(tads_frags$TAD_index)) {
  tad_frags=tads_frags[tads_frags$TAD_index==i,]
  vals<-approx(x=tad_frags$pos_TAD,y=tad_frags$scaled_DI,xout=fracs)
  tad_fracs<-cbind(rep(i,length(fracs)),vals$x,vals$y)
  tads_DI_scaled<-rbind(tads_DI_scaled,tad_fracs)
}
tads_DI_scaled<-tads_DI_scaled[2:nrow(tads_DI_scaled),]
colnames(tads_DI_scaled)<-c("TAD_index","TAD_pos","DI")
tads_DI_scaled<-as.data.frame(tads_DI_scaled,stringsAsFactors = FALSE)

#Plot Heatmap
tads_DI_scaled<-tads_DI_scaled[!(is.na(tads_DI_scaled$DI)),]
tads_DI_scaled$TAD_index<-as.character(tads_DI_scaled$TAD_index)
tads_DI_scaled$DI[tads_DI_scaled$DI>=2]<-2
tads_DI_scaled$DI[tads_DI_scaled$DI<=-4]<--4
ggplot(tads_DI_scaled)+
  geom_tile(aes(x=TAD_pos,y=TAD_index,fill=DI,color=DI),size=0.01)+
  scale_fill_gradient2(low="blue",high="red",mid="white",midpoint = 0,na.value="white",
   name="Directionality Index\n(normalized)")+
  scale_color_gradient2(low="blue",high="red",mid="white",midpoint = 0,na.value="white",
   name="")+
  theme_minimal()+
  guides(color=FALSE)+
  ggtitle("Directionality of selected TADs")+
  scale_x_continuous(name="TAD position")+
  scale_y_discrete(name="TAD")+
  theme(axis.text.y=element_blank())

```

# Fragment selection

When I digested the genome, I kept only fragments with a size from 150 to 800bp.

To select fragments, I took the 25 fragments with the lowest DI per TAD (which should represent one end of the TAD) and the 25 fragments with the highest DI per TAD (which should represent the other end of the TAD). If the top 25 negative fragments included positive fragments, I threw out the positive fragments, and vice versa.

### Fragment selection: DI of fragments.

Let's see how the DI of selected fragments compares to the DI of all fragments overlapping selected TADs:

```{r fragment_DI, echo=FALSE}
fragments<-fread("results/top_frags_filtered_tads.bed")
colnames(fragments)<-c("chr","start","stop","DI","index")

all_frags$type<-"Background"
all_frags$type[all_frags$start %in% fragments$start & all_frags$stop %in% fragments$stop]<-"fragment"

ggplot()+
  geom_histogram(aes(x=all_frags$DI,fill=all_frags$type),bins=50)+
  scale_x_continuous(name="Directionality Index",limits=c(-7,7))+
  scale_fill_manual(values=c("steelblue3","darkseagreen3"))+
  theme_minimal()+
  theme(legend.title=element_blank())+
  ggtitle("Directionality Index of Selected Fragments")
```

### Fragment selection: fragment per TAD

For each TAD, I selected the top 15 fragments with high DI, and the top 16 fragments with low DI. I then removed fragments from the high-DI group where the mean DI was negative, and vice versa. I removed all fragments from TADs with <32 fragments, as these TADs often had poor balance of directionality on either side and may have been mis-called.

Where do the selected fragments fall within the TADs?

```{r fragment_locs,echo=FALSE}
TADs_fragments<-matrix(ncol=6)
for (i in 1:nrow(tads)) {
  #Write out TAD quantiles
  chrs=paste("chr",rep(tads$V1[i],100),sep='')
  seqs<-seq(tads$V2[i],tads$V3[i],length.out=201)
  starts<-floor(seqs[1:200])
  stops<-floor(seqs[2:201])-1
  quants<-as.data.frame(cbind(chrs,starts,stops),stringsAsFactors = FALSE)
  write.table(quants,"temp_TAD.bed",quote=FALSE,sep='\t',row.names=FALSE,col.names=FALSE)  
  
  #Write out fragments belonging to TAD
  write.table(fragments[fragments$chr==paste("chr",tads$V1[i],sep='') & fragments$stop>=tads$V2[i] & fragments$start<=tads$V3[i],],"fragments_TAD.bed",quote=FALSE,sep='\t',row.names=FALSE,col.names=FALSE)  
  overlaps<-system("bedtools intersect -wa -a temp_TAD.bed -b fragments_TAD.bed",intern=TRUE)
  overlaps<-data.frame(do.call('rbind',strsplit(as.data.frame(overlaps,stringsAsFactors=FALSE)$overlaps,'\t',fixed=TRUE)))
  if (nrow(overlaps)>0) {
    #Find TAD quantiles with fragments
    colnames(overlaps)<-c("chr","starts","stops")
    system("rm temp_TAD.bed fragments_TAD.bed")
  
    # Label the quadrants by whether they have a fragment
    quants$type<-"No fragment"
    quants$type[quants$starts %in% overlaps$starts & quants$stops %in% overlaps$stops]<-"fragment"
    quants$index<-paste(tads$V1[i],tads$V2[i],tads$V3[i],sep='-')
    quants$quant<-1:nrow(quants)
    TADs_fragments<-rbind(TADs_fragments,as.matrix(quants))
  }
}
TADs_fragments<-as.data.frame(TADs_fragments,stringsAsFactors = FALSE)
TADs_fragments<-TADs_fragments[2:nrow(TADs_fragments),]
colnames(TADs_fragments)<-c("chr","start","stop","fragment","index","quant")
write.table(TADs_fragments,"TADs_fragments_quantile.txt",sep='\t',quote=FALSE,row.names=FALSE,col.names=FALSE)
TADs_fragments<-as.data.frame(fread("TADs_fragments_quantile.txt"),stringsAsFactors=FALSE)
colnames(TADs_fragments)<-c("chr","start","stop","fragment","index","quant")

TADs_fragments$fragment_bin[TADs_fragments$fragment=="No fragment"]<-0
TADs_fragments$fragment_bin[TADs_fragments$fragment=="fragment"]<-1
TADs_fragments$quant<-as.numeric(TADs_fragments$quant)

ggplot(TADs_fragments)+
  geom_tile(aes(y=index,x=quant,fill=fragment_bin,color=fragment_bin),size=0.01)+
  scale_fill_gradient(low="white",high="black",na.value="white",name="fragment")+
  scale_color_gradient(low="white",high="black",na.value="white",name="")+
  guides(color=FALSE)+
  theme_minimal()+
  ggtitle("fragment locations within selected TADs")+
  scale_x_continuous(name="TAD position")+
  scale_y_discrete(name="TAD")+
  theme(axis.text.y=element_blank(),
        legend.position="none")
```

# Probe selection

Ideally, I would get a probe from each selected fragment. However, there are two considerations:
1. Mapping: need probes that will not map to multiple regions
2. PCR amplification: need probes that won't induce amplification biases (due to high GC content)

Therefore, I'm selecting probes and performing filtering, based on the filtering that Noboru did when selecting probes for promoter capture HiC. Specifically, I: 
1. For each fragment, select 120bp probes from the beginning, middle, and end
2. Evaluate these probes for GC% and mappability
3. Iterate through the fragments, and keep the first probe, if any, that has GC between 25% and 60% and maps uniquely to the genome.

#### Length of probes: 120bp


```{r probeset_stats, echo=FALSE}
all_probes<-fread("results/candidate_probes.txt")
print(paste("Percent uniquely mapped, probes at beginning of fragments:",round(sum(all_probes$mapped1!="0")/nrow(all_probes),5)))
ggplot()+
  geom_histogram(aes(x=all_probes$GC1),bins=50)+
  theme_minimal()+
  ggtitle("GC content, probes at beginning of fragments")

print(paste("Percent uniquely mapped, probes at middle of fragments:",round(sum(all_probes$mapped2!="0")/nrow(all_probes),5)))
ggplot()+
  geom_histogram(aes(x=all_probes$GC2),bins=50)+
  theme_minimal()+
  ggtitle("GC content, probes at middle of fragments")

print(paste("Percent uniquely mapped, probes at end of fragments:",round(sum(all_probes$mapped3!="0")/nrow(all_probes),5)))
ggplot()+
  geom_histogram(aes(x=all_probes$GC3),bins=50)+
  theme_minimal()+
  ggtitle("GC content, probes at end of fragments")

print(paste("Percent of fragments surviving probe selection:",round(sum(!(is.na(all_probes$probeseq)))/nrow(all_probes),5)*100,"%",sep=''))

probes<-fread("results/selected_probes.txt")
print(paste("Number of probes: ",nrow(probes),sep=''))
print("Probe locations in frag:")
table(probes$probefragloc)
```


# Final TAD representation

### TADs by chromosome

Did I get a decent number of TADs per chromosome? Does the TAD calling generally reflect chromosome size?

Here, I'm looking at TADs after filtering: that is, after removing TADs<100kb, and removing TADs where there weren't enough good fragments or probes I couldn't find enough probes.


``` {r TADs_chr,echo=FALSE}
filtered_tads<-fread("results/filtered_tads.bed")
selected_tads<-fread("results/selected_tads.bed")
selected_tads$chr<-factor(selected_tads$V1,levels=c(1:22,"X","Y"))
ggplot()+
  geom_bar(aes(x=selected_tads$chr),fill="steelblue4")+
  annotate(x=13,y=200,"text",label=paste("Number of TADs:\n",nrow(selected_tads),sep=''))+
  scale_x_discrete(name="Chromosome")+
  theme_minimal()+
  ggtitle("TADs by chromosome after filtering")
```

### TAD filtering

How many TADs have fragments after filtering fragments for GC content and mappability? Am I still catching most TADs?
```{r TADs_with_frags, echo=FALSE}
print(paste("Total number of called TADs: ",nrow(tads),sep=''))

print(paste("Number of TADs > 100kb: ",nrow(tads[(tads$V3-tads$V2)>=100000,]),sep=''))

print(paste("Number of TADs > 100kb and with >=16 positive DI and >=16 negative DI fragments: ",nrow(filtered_tads),sep=''))

print(paste("Number of TADs > 100kb, with >=16 positive DI and >=16 negative DI fragments, and with >5 valid probes with positive and negative DI (final set):",nrow(selected_tads),sep=''))



      #How many tads after each filtering step: removing 100kb tads, removing tads with bad frags, removing tads with not enough probes
```


